content/blog_posts/articles/use-tsx-for-your-template-engine.md

```diff
@@ -0,0 +1,208 @@
+
+We love Edge and working with it, but it brings a few caveats.
+
+It is a "custom language". We need to have an extension for any IDE to have syntax highlighting. If we want to have type safety, we would need to create a complete LSP ([Language Server Protocol](https://en.wikipedia.org/wiki/Language_Server_Protocol)) from scratch, which is out of the scope of this project. It also means `prettier` does not work out of the box with Edge; a custom plugin would be needed.
+
+Those caveats may not be an issue for you, and you love working with Edge, but let me present you another alternative for those who would like better type-safety and IDE support.
+
+## JSX
+
+JSX (JavaScript Syntax Extension) is an XML-like syntax extension to ECMAScript without any defined semantics developed by Facebook.
+
+You may have already used it in React, Vue, Solid, or other frontend frameworks. While most of its use cases are inside React, JSX is not tied to it. It has [a proper spec](https://facebook.github.io/jsx/) for anyone to create and use a parser.
+
+The great part of JSX is its support. All IDE supports JSX; it already has an LSP, a prettier support, and even TypeScript has backed-in support.
+
+It makes it a great candidate for a templating engine.
+
+## Examples
+
+For those who have yet to use JSX or TSX (JSX on TypeScript file), let me show you some examples and what it may look like in your code.
+
+First, everything is a component; you define them with a function, and the function returns JSX, which follows an HTML-like syntax.
+
+```tsx
+export function Home() {
+  return (
+    <div>
+      <h1>Hello World</h1>
+    </div>
+  )
+}
+```
+
+Since this code resides in a `tsx` file, it can execute any JavaScript function like a standard function.
+
+```tsx
+import { cva } from 'class-variance-authority'
+
+interface ButtonProps { /* ... */ }
+
+export function Button(props: ButtonProps) {
+  const { color, children, size = 'medium', ...extraProps } = props
+  
+  const buttonStyle = cva('button', { /* ... */ }
+  
+  return (
+    <button class={button({ color, size })} {...extraProps}>
+      {children}
+    </button>
+  )
+}
+```
+
+We use `cva` from the [class-variance-authority](https://cva.style/docs) package in this example.
+
+We can leverage the [Async Local Storage](https://docs.adonisjs.com/guides/async-local-storage#async-local-storage) of AdonisJS to access the `HttpContext` anywhere in your template.
+
+:::warn
+
+ We recommend doing props drilling since using ALS will create a performance hit for your application.
+
+:::
+
+```tsx
+export function Header() {
+  const { auth } = HttpContext.getOrFail()
+  
+  if (auth.user) {
+    return <header>Authenticated!</header>
+  }
+  
+  return <header>Please connect!</header>
+}
+```
+
+## Setting up TSX
+
+I have tried different packages to use TSX as a templating engine; [JSXTE](https://github.com/ncpa0/jsxte) and [Kita](https://github.com/kitajs/html) were the most advanced ones.
```

Comment:
maybe this is out of scope for this article so feel free to ignore

but perhaps it would be good to include a very concise comparison between the two
from memory with kitajs you have to do props drilling or use ALS
whereas the other supports context

so yeah maybe add the most impactful props/cons for both tools

Reply:
I would prefer to remove the reference to `JSXTE` altogether. My goal was not to compare those, and it may create confusion. I will add some text to clarify that.